Nav/navigation md (#229)
software-mansion-labs/expensify-app-fork: contributingGuides/NAVIGATION.md

#### Navigating to a new screen

`Navigation.navigate` is used to go to a specific screen in the app or replace the existing one. This function works based on an extended implementation of the `linkTo` method. Because of this, there are some differences between using our function and `navigate` returned from `useNavigation` hook.

```ts
import Navigation from '@libs/Navigation/Navigation';
import ROUTES from '@src/ROUTES';
import interceptAnonymousUser from '@libs/interceptAnonymousUser';

// Basic navigation to a route
Navigation.navigate(ROUTES.HOME);

// Navigation with parameters
Navigation.navigate(ROUTES.SEARCH_ROOT.getRoute({
    query: 'type:expense status:all search',
    // additional parameters...
}));

// Navigation with forceReplace - replaces current screen instead of pushing a new one
Navigation.navigate(ROUTES.SEARCH_REPORT.getRoute({reportID: nextReportID, backTo}), {forceReplace: true});

// Navigation with a callback to handle anonymous users
interceptAnonymousUser(() => {
    Navigation.navigate(ROUTES.SETTINGS);
});
```

> [!NOTE]
> The most relevant differences between our implementation of `Navigation.navigate` and `navigate` returned from `useNavigation`:

> > 1. We pass `route` instead of a screen name, because our implementation of `Navigation.navigate` is based on `linkTo` method which accepts `route` as a parameter

Reviewer: I am wondering if mentioning the linkTo function is relevant at all

> 2. We import Navigation object from `@libs/Navigation/Navigation` not get it from the hook
> 3. Our method uses `PUSH` not `NAVIGATE` by default!!
> 4. We do not have a separate function `REPLACE`, to use this method you need to pass the `forceReplace` option to `Navigation.navigate`.

#### Going back

To navigate back, we use the `Navigation.goBack` function. We can call this function without any parameters, but the most common case is to call it with `backToRoute`. It is worth remembering that it is possible to deep link to any page in the application, and then we may lose a page to which we should go back from the navigation state. In such a case, we can simply use the mentioned parameter to indicate which page should be opened when going back. 

> [!NOTE]

> > This function should be used mainly with `backToRoute` param. If you want to use it, make sure there is a screen to which you should always go back in a given case and pass its route as a param.

Author: We want to strongly emphasize that this is the correct way to use `Navigation.goBack`

Let's consider the case when we have 4 routes on stack, 3 reports with different ids and search on the top. The diagram shows how navigating back works with and without parameter comparison.

```ts
import Navigation from '@libs/Navigation/Navigation';
import ROUTES from '@src/ROUTES';

Navigation.goBack(ROUTES.REPORT_WITH_ID.getRoute('1'), {compareParams: false});
```

```mermaid
flowchart TD
 subgraph subGraph0["compareParams = true"]
        B2["r/3"]
        A2["r/2"]
        C2["search"]
        D2["r/1"]
  end
 subgraph subGraph1["compareParams = false"]
        F2["r/1"]
        E2["search"]
        G2["r/2"]
        H2["r/3"]
  end
    B2 --> A2
    C2 L_C2_B2_0@-."Navigation.goBack('r/3')" .-> B2
    D2 --> C2
    F2 --> E2
    G2 --> F2
    H2 --> G2
    E2 L_E2_F2_2@-. "Navigation.goBack('r/3', {compareParams:false})" .-> F2
    A2 --> D2

    L_C2_B2_0@{ animation: fast } 
    L_E2_F2_2@{ animation: fast } 

```

#### Dismissing modals

> Many flows in an application are displayed in modals, one of them is RightModalNavigator (RHP) - the most commonly used modal in Expensify. This method is used to close it along with all open pages.

Reviewer 2: ```suggestion
Many flows in the application are displayed in modals. One such example is RightModalNavigator (RHP) - the most commonly used modal in Expensify. This method is used to close it along with all open pages.
```

```ts
import Navigation from '@libs/Navigation/Navigation';

// Simple modal dismissal
Navigation.dismissModal();
```

> [!NOTE]
> Why do we need a method other than `Navigation.goBack` to close modals? 
> Let's consider the following case: 
> You are going through a flow which has multiple steps. During this flow you want to close the entire modal, no matter which page you are on. If it was the first screen in this flow, there would be no difference between `Navigation.dismissModal` and `Navigation.goBack`. But after opening several pages in RHP, it is necessary to pop the entire RHP with all the open screens from the navigation state. This is exactly what `Navigation.dismissModal` does.

#### Dismissing modals with opening a report

> Another common case in Expensify is opening a report from RHP. This happens, for example, after creating a new expense. In this case, you need to close the modal and go to the report. So, as you might expect, you need to call `Navigation.dismissModal` and then `Navigation.navigate` to the report, or even just `Navigation.navigate` will be sufficient if you do not want close the modal and be able to go back to it. And that's basically true, but there are a few additional aspects we need to pay attention to and they are covered in `dismissModalWithReport`.

Reviewer 2: ```suggestion
This happens, for example, after creating a new expense. In this case, you need to close the modal and then go to the report. So, as you might expect, you need to call `Navigation.dismissModal` and then `Navigation.navigate` (after that?) [...]
```

More generally I believe this part:
```
So, as you might expect, you need to call `Navigation.dismissModal` and then `Navigation.navigate` to the report, or even just `Navigation.navigate` will be sufficient if you do not want close the modal and be able to go back to it.
```
Is too long of a sentence. Break it down.

> 2. On a wide screen we need to be sure that the modal has been closed, only then we want to navigate to the report. For this purpose, `navigate` called after `dismissModal` is wrapped in `InteractionManager.runAfterInteractions`.

#### Summary

- `Navigation.navigate` is used to navigate between screens. Remember that it calls the `linkTo` method implemented by us. It accepts the route as a parameter not a screen name.
- `Navigation.goBack` allows you to navigate back to the previous page. You will probably need to pass the `backToRoute` parameter to this method in such cases to preserve the screen order after refreshing the page.
- If you want to go back to the screen regardless of its parameter values, pass `{compareParams: false}` to `Navigation.goBack`.
- If you want to close the entire modal window, regardless of how many pages you have opened, use `Navigation.dismissModal` to do that.
- If you want to open a report from RHP to prevent navigation back to this modal window, use `Navigation.dismissModalWithReport`.

### Adding new screens

> To add a new screen you need to:

Reviewer 2: maybe a new heading for this part as well?

    const SETTINGS_TO_RHP: Partial<Record<keyof SettingsSplitNavigatorParamList, string[]>> = {
        // ...existing code
        [SCREENS.SETTINGS.PREFERENCES.ROOT]: [
            // ...existing code
            SCREENS.SETTINGS.NEW_SCREEN,
        ],
    };

    export default SETTINGS_TO_RHP;
    ```

### Debugging

#### Reading state when it changes

Often, to find the cause of a bug, it is worth checking how the state changes. To do that quickly, add `console.log` to the `handleStateChange` method in `NavigationRoot.tsx`

```ts
// src/libs/Navigation/NavigationRoot.tsx

    const handleStateChange = (state: NavigationState | undefined) => {
        //...existing code
        // Add console.log here to read the updated state
        console.log("state", state);
    };
```

#### Finding the code that calls the navigation function

The easiest way to find the piece of code from which the navigation method was called is to use a debugger and breakpoints. You should attach a breakpoint in the navigation method and check the call stack, this way you can easily find the navigation method that caused the problem.

### Using `backTo` route param

Let's consider the following case: 

> One screen is accessible from two or more different screens. We need to remember that we can pass `backToRoute` to `Navigation.goBack`, but now it should be set dynamically depending on the screen from which we opened the current page. 

Reviewer 2: ```suggestion
One screen is accessible from two or more different screens.
```

I don't think it's obvious what "accessible" means in this context. Does it mean that you _can_ navigate from 2 different screens into another screen?
In general I would expect every that one can navigate to any Screen from any other screen (or place) in the App.
So maybe some different wording here?

    * Builds a URL with an encoded URI component for the `backTo` param which can be added to the end of URLs
    */
    function getUrlWithBackToParam<TUrl extends string>(url: TUrl, backTo?: string, shouldEncodeURIComponent = true): `${TUrl}` {
        const backToParam = backTo ? (`${url.includes('?') ? '&' : '?'}backTo=${shouldEncodeURIComponent ? encodeURIComponent(backTo) : backTo}` as const) : '';
        return `${url}${backToParam}` as `${TUrl}`;
    }

    const ROUTES = {
        NEW_SETTINGS_SCREEN_WITH_BACK_TO: {
            route: 'new-settings-screen',
            getRoute: (backTo?: string) => getUrlWithBackToParam('new-settings-screen', backTo),
        },
    } as const;
    ```

2. Find the screen to which you want to add `backTo` parameter in `src/libs/Navigation/types.ts` and define it:

```ts
type SettingsSplitNavigatorParamList = {
    ...existing types
    [SCREENS.SETTINGS.NEW_SCREEN_WITH_BACK_TO]: {backTo?: string}
};
```

1. When navigating to this screen from a non-default screen, pass a route to `Navigation.navigation` with `backTo` parameter.

```ts
Navigation.navigate(ROUTES.NEW_SETTINGS_SCREEN_WITH_BACK_TO.getRoute(Navigation.getActiveRoute()))

> ```

Reviewer 2: The item above say that we need to pass the `backTo` param.
So aren't we missing a `backTo` param being passed in this code snippet?

Reviewer 2: ah shit, my bad I misread this line

4. In the new screen read `backTo` from `route.params` and pass it to `Navigation.goBack`.

```ts
// src/pages/NewSettingsScreen.tsx
import type {SettingsSplitNavigatorParamList} from '@libs/Navigation/types';
import type {PlatformStackScreenProps} from '@libs/Navigation/PlatformStackNavigation/types';
import HeaderWithBackButton from '@components/HeaderWithBackButton';
import ScreenWrapper from '@components/ScreenWrapper';
import Navigation from '@libs/Navigation/Navigation';
import ROUTES from '@src/ROUTES';

type NewSettingsScreenNavigationProps = PlatformStackScreenProps<SettingsSplitNavigatorParamList, typeof SCREENS.SETTINGS.NEW_SCREEN_WITH_BACK_TO>;

const DEFAULT_ROUTE_TO_GO_BACK = ROUTES.SETTINGS;

function NewSettingsScreen({route}: NewSettingsScreenNavigationProps) {
    // Your screen component code here

    return (
        <ScreenWrapper>
          <HeaderWithBackButton
                onBackButtonPress={() => Navigation.goBack(route.params?.backTo ?? DEFAULT_ROUTE_TO_GO_BACK)}
            />
            {/* Your screen content */}
        </ScreenWrapper>
    );
}
```

> ### Genereting state from a path

Reviewer 2: typo Genereting

> In the following section you will find information on how the navigation state is generated from the path.

Reviewer 2: [nitpick] maybe instead of "from the path" we say from `route` or from `url`?

                "key": "stack-NtlStqLSBYzpsUbgmVOFe",
                "index": 0,
                "routes": [
                    {
                        "name": "Settings",
                        "state": {
                            "stale": false,
                            "type": "stack",
                            "key": "stack-Y5L11fQW5t82MpOEZEUis",
                            "index": 0,
                            "routes": [
                                {
                                    "name": "Settings_Display_Name",
                                    "path": "/settings/profile/display-name",
                                    "key": "Settings_Display_Name-G0pJsgF8dTHLlbzVXontn"
                                }
                            ],
                        },
                        "key": "Settings-cy2uuw6pJmxswxNpRUvl2"
                    }
                ],
            },
            "key": "RightModalNavigator-b1XDwV6vIF23wu4bITfXl"
        }
    ],
}
```

In the above example, we can see that when building a state from a link leading to a screen in RHP, screens that appear below the overlay are also built.

> ### Setting the correct screen below RHP

Reviewer 2: maybe [nitpick] but to me "below" might mean:
```
top of something
A
B
C <--- C is below B
```

and maybe "underneath" would be better?

> RHP screens can usually be opened from a specific central screen, of course there are cases where one RHP screen can be used in different tabs (then using `backTo` parameter proves useful), however most often one RHP screen has a specific central screen assigned underneath.

Reviewer 2: [nitpick] too long of a sentence IMO, could be like 3 separate ones.

- `src/libs/Navigation/AppNavigator/createSplitStackNavigator/index.tsx`

```ts
function useCustomEffects(props: CustomEffectsHookProps) {
    useNavigationResetOnLayoutChange(props);
    usePreserveNavigatorState(props.state, props.parentRoute);
}

function useCustomSplitNavigatorState({state}: CustomStateHookProps) {
    const {shouldUseNarrowLayout} = useResponsiveLayout();

    const sidebarScreenRoute = state.routes.at(0);

    if (!sidebarScreenRoute) {
        return state;
    }

    const centralScreenRoutes = state.routes.slice(1);
    const routesToRender = shouldUseNarrowLayout ? state.routes.slice(-2) : [sidebarScreenRoute, ...centralScreenRoutes.slice(-2)];

    return {...state, routes: routesToRender, index: routesToRender.length - 1};
}
```
In `SplitNavigators`, only the last 2 routes are rendered in a similar way, but we have to also ensure that the sidebar is on the wide screen at 0 index in the state of this navigator.

> [!NOTE]
> When nested routes are not rendered their state is lost and when returning to these screens it has to be recreated. To do this the state is saved in the `preservedNavigatorStates` object using the `usePreserveNavigatorState` hook.

> ### State persistance after page refresh

Reviewer 2: typo - persistence

                    },
                });
            }
            return;
        }

        Navigation.navigate(ROUTES.WORKSPACES_LIST.route);
    });
}, [shouldUseNarrowLayout]);
```

## Custom navigators

`react-navigation` provides a few fundamental navigators: `StackNavigator`, `DrawerNavigator` etc.

To handle unique application requirements, custom navigators were created to help extend the functionality of the basic `StackNavigator`: `RootStackNavigator`, `SplitNavigator` and `SearchFullscreenNavigator`.

### RootStackNavigator

This type of navigator is used only once in the app and it is in the root of navigation. It extends the basic functionality of `StackNavigator`. 
Custom functionalities handled by this navigator:
- dismissing modals,
- animating central screens on a narrow layout,
- preventing users from going back in history during onboarding.

### FullscreenNavigator / SplitNavigator

These navigators cover the entire screen and do not have transparent overlay. Each of them has a navigation tab bar icon that is highlighted when that screen is at the top of the navigation stack or visible under the mdal navigator overlay.

> It is worth noting that despite the navigation tab is displayed, the application does not use `BottomTabNavigator`. When one of the navigation tab bar buttons is pressed, a new full screen is pushed onto the root stack. `StackNavigator` was used to implement this groups of screens as it has more flexibility and preserves navigation history in the browser.

Reviewer 2: ```
It is worth noting that despite the navigation tab is displayed, the application does not use `BottomTabNavigator`. 
```

Something is no-yes with this sentence ;)

A subset of `FullScreenNavigators` are `SplitNavigators`:

This type of navigator is also based on `StackNavigator`. It has two types of screens.

1. ***Sidebar screen:***
   There is only one screen of this type on the stack and it is always the first one in the SplitNavigator stack (it is present at 0 index in SplitNavigator's routes).
2. ***Central screen:***
   All other screens displayed next to the sidebar.

On a narrow layout it behaves as a normal `StackNavigator`.

On a wide layout, a sidebar screen is translated to the left to make it visible. This way, the user will see both a sidebar screen and a central screen.

Thanks to this navigator, there is always at least one side screen and one center screen in a wide layout.

> [!NOTE]
> To check if a route belongs to a specific group, it is worth using the functions available in `@libs/Navigation/helpers/isNavigatorName`. For example, to check if a route is a `FullScreenNavigator` you can use the `isFullScreenName` function.

> #### When we need to adapt the split navigator state (function `adaptStateIfNecessary`)

Reviewer 2: not sure if I'm correct, but should this be: 
```
When >do< we need to adapt the [...]
```
?

The purpose of `adaptStateIfNecessary` function is to ensure that a given `SplitNavigator` has a sidebar and a central screen when necessary.

 When is this function used:
- This function is called when the application starts . If we open the application on the central screen, this function will push the sidebar screen.
- When we are on the sidebar on a small screen and we expand it to a wide layout we have to push the central screen to fill the space on the screen.

> [!NOTE]
> `adaptStateIfNecessary` is called in the `getInitialState` and `getRehydratedState` methods in `src/libs/Navigation/AppNavigator/createSplitNavigator/SplitRouter.ts`.

`FullScreenNavigators` in the app:

#### `NAVIGATORS.SEARCH_FULLSCREEN_NAVIGATOR` (Reports tab) 

Something worth noticing is even though the Search pages may visually look like a split navigator, it is `FullScreenNavigator` with additional `ExtraContent` which displays `<SearchSidebar />`. It is implemented this way to meet the requirement that the sidebar and the central screen of the Search page have the same URL. 

#### `NAVIGATORS.REPORTS_SPLIT_NAVIGATOR` (Inbox tab)

> It includes the `HOME` screen (`<BaseSidebarScreen />` component) with a list of reports as a sidebar screen and the `REPORT` screen displayed as a central one. There can be multiple report screens in the stack with different report IDs.

Reviewer 2: ```suggestion
It displays the `HOME` screen (`<BaseSidebarScreen />` component) with a list of reports as a sidebar screen and the `REPORT` screen displayed as a central one. There can be multiple report screens in the stack with different report IDs.
```

#### `NAVIGATORS.SETTINGS_SPLIT_NAVIGATOR` (Account tab)

`SettingsSplitNavigator` is responsible for displaying user profile settings screens. The URLs of these pages start with `/settings` and the sidebar component is `<InitialSettingsPage />`. 

#### `NAVIGATORS.WORKSPACE_SPLIT_NAVIGATOR` (Workspaces tab)

`WorkspaceSplitNavigator` displays the settings of the selected workspace (the URLs start with `/settings/workspaces`). `<WorkspaceInitialPage />` is the sidebar screen component of this navigator.

> [!NOTE]
> The Workspaces tab is also selected when the workspace list is displayed (`SCREENS.WORKSPACES_LIST`) which is a separate screen displayed in `AuthScreens` next to the other navigators!

### Modals

These screens / navigators have a transparent overlay underneath.

> On wide layout we have functionality that ensures that there is at least one full screen under the modal on the stack that will be visible under overlay.

Reviewer 2: ```
[...] that ensures that there is at least one full screen under the modal >>on the stack that will be visible<< under overlay.
```

Kinda reads like it will be **the stack** that will be visible under. I would rephrase this somehow.

    [SCREENS.SETTINGS.PROFILE.ROOT]: [
        // ... existing screens ...
        SCREENS.SETTINGS.NEW_RHP_SCREEN, // Add your new screen here
    ],
    // ... rest of the mapping ...
};
```

> [!NOTE]
> If the RHP screen is not added to any relation, the Inbox tab will be opened underneath by default.

> [!NOTE]
> If a given RHP has a route param `backTo`, the relation of the screen passed as `backTo` will be checked. This allows reusing RHP screens in different tabs.

## Performance solutions

Navigation-related performance improvements have been implemented in the following files:

-   `src/libs/Navigation/AppNavigator/createRootStackNavigator/index.tsx`

The optimization is designed to improve performance by limiting the number of mounted screens in the navigation stack.

```ts
function useCustomRootStackNavigatorState({state}: CustomStateHookProps) {
    const lastSplitIndex = state.routes.findLastIndex((route) => isFullScreenName(route.name));
    const routesToRender = state.routes.slice(Math.max(0, lastSplitIndex - 1), state.routes.length);

    return {...state, routes: routesToRender, index: routesToRender.length - 1};
}
```

> To optimize the number of routes rendered in `RootStackNavigator` we limit the number of `FullScreenNavigators` rendered to 2.

Reviewer: Maybe it's worth mentioning that we use last two instead of just one to make sure animations works properly
